### PR TITLE
PyO3 & Scikit introduction

### PR DESCRIPTION
**Pull request by levkk**
1. Add sklearn wrappers.
2. Replace smartcore logistic with scikit temporarily as a POC.

### CODE REVIEW

File: pgml-extension/pgml_rust/sql/schema.sql [hunk 1]
```diff
@@ -77,6 +77,7 @@ CREATE TABLE IF NOT EXISTS pgml_rust.models(
 	project_id BIGINT NOT NULL,
 	snapshot_id BIGINT NOT NULL,
 	algorithm TEXT NOT NULL,
+	backend TEXT DEFAULT 'smartcore',
```
Comment (montanalow) [inline pgml-extension/pgml_rust/sql/schema.sql:80]: Should we really have a default here?

File: pgml-extension/pgml_rust/src/backends/backend.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+use pgx::*;
+use serde::Deserialize;
+
+#[derive(PostgresEnum, Copy, Clone, Eq, PartialEq, Debug, Deserialize)]
+#[allow(non_camel_case_types)]
+pub enum Backend {
+    xgboost,
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/backends/backend.rs:7]: Some of these backends don't differentiate between python/rust implementations, and seem to be confounded with the algorithm.

File: pgml-extension/pgml_rust/src/backends/backend.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+use pgx::*;
+use serde::Deserialize;
+
+#[derive(PostgresEnum, Copy, Clone, Eq, PartialEq, Debug, Deserialize)]
+#[allow(non_camel_case_types)]
+pub enum Backend {
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/backends/backend.rs:6]: Backend is a heavily overloaded term. I'd prefer something like `library`, `implementation`, `engine`, `provider`...

File: pgml-extension/pgml_rust/src/backends/sklearn.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    });
+
+    version
+}
+
+pub fn sklearn_train(
+    algorithm_name: &str,
+    dataset: &Dataset,
+    hyperparams: HashMap<String, f32>,
+) -> SklearnBox {
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/backends/sklearn.rs:26]: This is pretty amazing.

File: pgml-extension/pgml_rust/src/backends/sklearn.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                        hyperparams.into_py(py),
+                    ],
+                ),
+            )
+            .unwrap();
+
+        train
+            .call1(
+                py,
+                PyTuple::new(py, &[dataset.x_train(), dataset.y_train()]),
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/backends/sklearn.rs:53]: It's crazy that this works.

File: pgml-extension/pgml_rust/src/backends/wrappers.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+def predictor(estimator, num_features):
+    return predictor_joint(estimator, num_features, 1)
+
+
+def predictor_joint(estimator, num_features, num_targets):
+    def predict(X):
+        X = np.asarray(X).reshape((-1, num_features))
+        y_hat = estimator.predict(X)
+
+        # Only support single value models for just now.
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/backends/wrappers.py:94]: Can you make this a TODO, so it's easier to grep?

File: pgml-extension/pgml_rust/src/orm/estimator.rs [hunk 7]
```diff
@@ -431,3 +432,82 @@ impl Estimator for BoosterBox {
...
 }
+
+/// A wrapper around a Scikit estimator.
+/// The estimator is a Python object and can only be used
+/// inside Python::with_gil.
+pub struct SklearnBox {
+    pub contents: Box<Py<PyAny>>,
+}
+
+impl Drop for SklearnBox {
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/orm/estimator.rs:443]: Why do you need to implement this manually? Looks like fun...

File: pgml-extension/pgml_rust/src/orm/estimator.rs [hunk 8]
```diff
@@ -431,3 +432,82 @@ impl Estimator for BoosterBox {
...
+        formatter.debug_struct("SklearnBox").finish()
+    }
+}
+
+impl serde::Serialize for SklearnBox {
+    fn serialize<S>(&self, _serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        unreachable!()
```
Comment (montanalow) [inline pgml-extension/pgml_rust/src/orm/estimator.rs:490]: should this be `todo!()` or is it really unreachable!()? Maybe we should modify the upstream interface, I mean trait definition...

File: pgml-extension/pgml_rust/src/orm/estimator.rs (hunk 7 above)
Comment (levkk) [reply pgml-extension/pgml_rust/src/orm/estimator.rs:443]: Just GIL things.

File: pgml-extension/pgml_rust/src/backends/backend.rs (hunk 2 above)
Comment (levkk) [reply pgml-extension/pgml_rust/src/backends/backend.rs:7]: You can pickle an xgboost class (and we do in 1.0). Native xgboost is serialized differently.

File: pgml-extension/pgml_rust/sql/schema.sql (hunk 1 above)
Comment (levkk) [reply pgml-extension/pgml_rust/sql/schema.sql:80]: Not really.

File: pgml-extension/pgml_rust/src/backends/sklearn.rs (hunk 5 above)
Comment (levkk) [reply pgml-extension/pgml_rust/src/backends/sklearn.rs:53]: It's science

File: pgml-extension/pgml_rust/src/backends/wrappers.py (hunk 6 above)
Comment (levkk) [reply pgml-extension/pgml_rust/src/backends/wrappers.py:94]: It's in my head don't worry

File: pgml-extension/pgml_rust/src/orm/estimator.rs (hunk 8 above)
Comment (levkk) [reply pgml-extension/pgml_rust/src/orm/estimator.rs:490]: Definitely maybe a todo

**Pull request merged**